### PR TITLE
[RFC] Introduce K_SYSCALL_INLINE

### PR DESCRIPTION
**Pull request by gmarull**
Add a new utility macro to define inline syscalls. This is an
alternative to the existing way of defining inline syscalls, ie

```c
K_SYSCALL_INLINE(int, foo, int bar)
{
    ...
}
```

instead of:

```c
__syscall int foo(int bar);

static inline int z_impl_foo(int bar)
{
    ...
}
```

This solution has several advantages:

- Less verbose
- Non-breaking change, can co-exist with current solution
- Hides implementation details (ie, z_impl_...) from header files
- Opens the door to skip the generation of syscall files by just expanding `K_SYSCALL_INLINE()` to a single `static inline ...` call if `CONFIG_USERSPACE=n` (not done for now), thus speeding up builds. (not tested, but unless I've missed something this should be possible)

### CODE REVIEW

**Comment by dcpleung** [comment]
This makes grepping for `__syscall` and `z_impl_` impossible. Not to mention that you now have two parallel ways of defining syscalls, which adds to confusion to developers.

**Comment by gmarull** [comment]
> This makes grepping for `__syscall` and `z_impl_` impossible. 


You can grep for K_SYSCALL_INLINE

> Not to mention that you now have two parallel ways of defining syscalls, which adds to confusion to developers.

Both are kept to not break users, but the idea would be to move to the new method when the inline case applies, ie, the majority of device driver headers.

**Comment by dcpleung** [comment]
> > This makes grepping for `__syscall` and `z_impl_` impossible.
> You can grep for K_SYSCALL_INLINE

Now you have to grep twice to get them.

> > Not to mention that you now have two parallel ways of defining syscalls, which adds to confusion to developers.
> Both are kept to not break users, but the idea would be to move to the new method when the inline case applies, ie, the majority of device driver headers.

A lot of these are API declarations, where once written, hardly changes.
And yet, you are still introducing a parallel way to define syscalls and their implementations, as this cannot be used with `z_impl_` that are not inlined.

I am not sure about the benefits you have mentioned...
* Being less verbose is not necessarily a good thing. You hide the actual function declaration behind a macro. Now when developers are skimming through the header, they may no longer register those as functions that can be called. Since you have written these, you probably have your mind trained on these already. But for others, especially for those who are not maintaining APIs, this can be tiring to look at. Recognizing a new pattern requires some training on the brain.
* Currently, `gen_syscalls.py` does `static inline` for the syscall entry function, which simply calls the corresponding `z_impl_` function if userspace is not enabled.

I think the biggest questions are: how IDE/editors and doxygen react to this change. Not everyone is using IDE and editors which relies on compiling the code over and over again to grab function signatures (maybe ctags? not sure about that though). Will doxygen be able to expand the macros and see the functions so that it can extract the signatures and associate them with the correct doc blocks?

**Comment by nashif** [comment]
> I think the biggest questions are: how IDE/editors and doxygen react to this change. Not everyone is using IDE and editors which relies on compiling the code over and over again to grab function signatures (maybe ctags? not sure about that though). Will doxygen be able to expand the macros and see the functions so that it can extract the signatures and associate them with the correct doc blocks?

still struggling to get ZTEST test documentation to work as before, not it is completely unusable, would be great if there was a soluton.

**Comment by henrikbrixandersen** [comment]
> A lot of these are API declarations, where once written, hardly changes.
> And yet, you are still introducing a parallel way to define syscalls and their implementations, as this cannot be used with `z_impl_` that are not inlined.

This is my main concern with this proposal as well. I really don't think we need _more_ macros in Zephyr, especially not for code constructs that are already rather clear.

**Comment by fabiobaltieri** [comment]
Would a non-inline variant make possible to phase out the current schema? Like `K_SYSCALL_INLINE` and `K_SYSCALL`

**Comment by gmarull** [comment]
> > A lot of these are API declarations, where once written, hardly changes.
> > And yet, you are still introducing a parallel way to define syscalls and their implementations, as this cannot be used with `z_impl_` that are not inlined.
> 
> This is my main concern with this proposal as well. I really don't think we need _more_ macros in Zephyr, especially not for code constructs that are already rather clear.

This macro is 1 expansion away, unlike `DT_*` or some `util.h` craziness, or the ugliness of the stats macros in some driver class headers. I have to disagree that existing constructs are clear. Reading driver headers has become far from trivial from a newbie perspective.

With the proposal here, I'd also like to emphasize that it could open the door to skip the syscall file generation if not using userspace (we're now always parsing some headers at least twice, Python+Compiler, even if one doesn't care about userspace at all). It should be possible to opt-out, now it's not.

**Comment by gmarull** [comment]
> Would a non-inline variant make possible to phase out the current schema? Like `K_SYSCALL_INLINE` and `K_SYSCALL`. I like that this makes things more compact.

I guess it's possible. We could also introduce `DEVICE_SYSCALL*` and leave the rest as is, my main focus is to improve the user experience when writing drivers.

**Comment by dcpleung** [comment]
> > Would a non-inline variant make possible to phase out the current schema? Like `K_SYSCALL_INLINE` and `K_SYSCALL`. I like that this makes things more compact.
> 
> I guess it's possible. We could also introduce `DEVICE_SYSCALL*` and leave the rest as is, my main focus is to improve the user experience when writing drivers.

How would it improve the experience on writing individual drivers? Most of these syscalls are on the API level and not individual drivers. These is rarely any need to mess with them on a constant basis. I would argue that it makes the developer experience worse as IDEs would not be able to do proper syntax highlighting since these are no longer recognized as functions.

**Comment by gmarull** [comment]
> > > Would a non-inline variant make possible to phase out the current schema? Like `K_SYSCALL_INLINE` and `K_SYSCALL`. I like that this makes things more compact.
> > 
> > 
> > I guess it's possible. We could also introduce `DEVICE_SYSCALL*` and leave the rest as is, my main focus is to improve the user experience when writing drivers.
> 
> How would it improve the experience on writing individual drivers?

I meant `using`, my mistake.

**Comment by henrikbrixandersen** [comment]
> > How would it improve the experience on writing individual drivers?
> 
> I meant `using`, my mistake.

But the generated API documentation, which in my opinion is a very well-suited source of information for people _using_ drivers, stays the same with this change?

**Comment by gmarull** [comment]
> > > How would it improve the experience on writing individual drivers?
> > 
> > 
> > I meant `using`, my mistake.
> 
> But the generated API documentation, which in my opinion is a very well-suited source of information for people _using_ drivers, stays the same with this change?

it does, https://builds.zephyrproject.io/zephyr/pr/71402/docs/doxygen/html/group__can__interface.html#ga4af6d0d9ab72b195909f511ac65cb8fa

**Comment by henrikbrixandersen** [comment]
> it does, https://builds.zephyrproject.io/zephyr/pr/71402/docs/doxygen/html/group__can__interface.html#ga4af6d0d9ab72b195909f511ac65cb8fa

So how does this PR improve the experience for developers using drivers?

**Comment by gmarull** [comment]
> > it does, https://builds.zephyrproject.io/zephyr/pr/71402/docs/doxygen/html/group__can__interface.html#ga4af6d0d9ab72b195909f511ac65cb8fa
> 
> So how does this PR improve the experience for developers using drivers?

By providing less cluttered header files (e.g. with implementation details like `z_impl_*`), which is the entry point to an API. The current solution would likely allow to just expand to a simple `static inline` in case CONFIG_USERSPACE=n, making things even more clear if one uses a _modern_ IDE where expansions are shown e.g. on hover. I've always found syscalls really invasive (e.g. enforced parsing), and often one doesn't care about userspace.

File: include/zephyr/drivers/can.h [hunk 1]
```diff
@@ -825,9 +826,7 @@ struct can_device_state {
  *
  * @return 0 on success, or a negative error code on error
  */
-__syscall int can_get_core_clock(const struct device *dev, uint32_t *rate);
-
-static inline int z_impl_can_get_core_clock(const struct device *dev, uint32_t *rate)
```
Comment (nashif) [inline include/zephyr/drivers/can.h:830]: this works here where both the syscall and the implementation are in the same header, how is this going to work for something like https://github.com/zephyrproject-rtos/zephyr/blob/4a30aab2aefab891ba4b47ab2eed761c2210c53a/include/zephyr/kernel.h#L282?

Comment (gmarull) [reply include/zephyr/drivers/can.h:830]: This is not an _inline_ syscall, ie, z_impl_* inline below. A _dummy_ `K_SYSCALL` macro would be enough to have the same scheme, ie

```c
#define K_SYSCALL(ret, name, ...) static inline ret name(__VA_ARGS__)
```

**Comment by nashif** [comment]
> and often one doesn't care about userspace

We should not segment and change the zephyr code based on how much one cares for something or not, mostly based on vendors and their usecases. Because the minute you start doing this, I can come up with a long list of things I do not care about that are always in the way, sometime preventing me from doing the most basic things but I learned to live with those, knowing they are useful for someone else.

What we have here is nothing out of this world and is very common. Masking this with macros is not going to help IMO, instead, this could be documented better and users will need to deal with that. I personally did not see any complaints or issues of this model, we have many areas many users struggle with, this has not been on the top AFAIK.

**Comment by gmarull** [comment]
> > and often one doesn't care about userspace
> 
> We should not segment and change the zephyr code based on how much one cares for something or not, mostly based on vendors and their usecases. Because the minute you start doing this, I can come up with a long list of things I do not care about that are always in the way, sometime preventing me from doing the most basic things but I learned to live with those, knowing they are useful for someone else.
> 

This is not segmenting, just trying to be less intrusive. Userspace requires parsing all syscall include dirs/files regardless of userspace being enabled, for example. This should be an opt-out feature, IMHO.

> What we have here is nothing out of this world and is very common. Masking this with macros is not going to help IMO, instead, this could be documented better and users will need to deal with that. I personally did not see any complaints or issues of this model, we have many areas many users struggle with, this has not been on the top AFAIK.

File: include/zephyr/drivers/can.h (hunk 1 above)
Comment (nashif) [reply include/zephyr/drivers/can.h:830]: ok, and how is 

 `K_SYSCALL(int, k_thread_stack_free, k_thread_stack_t *stack);` 

better than plain and direct 

`__syscall int k_thread_stack_free(k_thread_stack_t *stack); `?

**Comment by dcpleung** [comment]
> > > it does, https://builds.zephyrproject.io/zephyr/pr/71402/docs/doxygen/html/group__can__interface.html#ga4af6d0d9ab72b195909f511ac65cb8fa
> > 
> > So how does this PR improve the experience for developers using drivers?
> 
> By providing less cluttered header files (e.g. with implementation details like `z_impl_*`), which is the entry point to an API. The current solution would likely allow to just expand to a simple `static inline` in case CONFIG_USERSPACE=n, making things even more clear if one uses a _modern_ IDE where expansions are shown e.g. on hover. I've always found syscalls really invasive (e.g. enforced parsing), and often one doesn't care about userspace.

I still fail to see why user of driver APIs care about `z_impl_*()` or the actual implementation unless they are debugging driver issues (and it's even worse as GDB does not expand macros). The API is the syscalls themselves without `z_impl_` prefix. It is even LESS beneficial to users of driver APIs as this hides the signature of the APIs themselves, simply because they would _**not**_ see the API functions but some macros.

Your point of this working with CONFIG_USERSPACE=n is moot here as this PR does not provide a full solution to that. Instead, that is just hand waving at this point. Besides, how do you explain to others why some syscalls have corresponding `z_impl_*()` and some don't? How do you plan to eventually tackle the issue of `z_impl_*()` in C source code outside of headers. How do you plan to deal with `z_vrfy_*()`? Right now, these is always the same way via a three part combo: API + `z_vrfy_` + `z_impl_`. With your proposal, now people have to understand what `K_SYSCALL_INLINE()` means, and where `z_impl_` is.

And you can also have so-called "decluttered" headers by grouping the API declarations together and their `z_impl_` counterparts into another groups, such that they do not interleave with each other.

And I **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** _**REALLY**_ need to make this one point very **very** _**very**_ abundantly clear, that you don't care about userspace does not mean that no one cares (or not enough people care). And to be very blunt, the union of other developer's working set is _not_ a subset of yours.

**Comment by gmarull** [comment]
> > > > it does, https://builds.zephyrproject.io/zephyr/pr/71402/docs/doxygen/html/group__can__interface.html#ga4af6d0d9ab72b195909f511ac65cb8fa
> > > 
> > > 
> > > So how does this PR improve the experience for developers using drivers?
> > 
> > 
> > By providing less cluttered header files (e.g. with implementation details like `z_impl_*`), which is the entry point to an API. The current solution would likely allow to just expand to a simple `static inline` in case CONFIG_USERSPACE=n, making things even more clear if one uses a _modern_ IDE where expansions are shown e.g. on hover. I've always found syscalls really invasive (e.g. enforced parsing), and often one doesn't care about userspace.
> 
> I still fail to see why user of driver APIs care about `z_impl_*()` or the actual implementation unless they are debugging driver issues (and it's even worse as GDB does not expand macros). The API is the syscalls themselves without `z_impl_` prefix. It is even LESS beneficial to users of driver APIs as this hides the signature of the APIs themselves, simply because they would _**not**_ see the API functions but some macros.
> 
> Your point of this working with CONFIG_USERSPACE=n is moot here as this PR does not provide a full solution to that. Instead, that is just hand waving at this point. Besides, how do you explain to others why some syscalls have corresponding `z_impl_*()` and some don't? How do you plan to eventually tackle the issue of `z_impl_*()` in C source code outside of headers. How do you plan to deal with `z_vrfy_*()`? Right now, these is always the same way via a three part combo: API + `z_vrfy_` + `z_impl_`. With your proposal, now people have to understand what `K_SYSCALL_INLINE()` means, and where `z_impl_` is.
> 
> And you can also have so-called "decluttered" headers by grouping the API declarations together and their `z_impl_` counterparts into another groups, such that they do not interleave with each other.
> 
> And I **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** **REALLY** _**REALLY**_ need to make this one point very **very** _**very**_ abundantly clear, that you don't care about userspace does not mean that no one cares (or not enough people care). And to be very blunt, the union of other developer's working set is _not_ a subset of yours.

I'm not sure guys from where you've got the idea that I do not care about userspace. I'm definitely not a userspace user, but wether I like it or not, I need to deal with it. This PR is just a discussion on how to improve things like:

- How to make driver API headers a bit more readable, less cluttered, without impl. details
- Open discussions on minimizing userspace impact on users that do not require it, for example, by not parsing almost all headers twice and autogenerating code for nothing (I want to believe this would not be accepted today).

Zephyr will never cease surprising me. Macro craziness in DT or util.h may be seen as "necessary" but then adding a 1-level expansion macro is seen as a "problem". Or some breaking changes may be seen as "unacceptable", but then things like HWMv2 which basically breaks almost every single application out there is "okay". I really miss a core team guiding this project, and someone who listens to users feedback.

**Comment by dcpleung** [comment]
Yet the answer to the question of how hiding function declarations would be beneficial to users of those functions is still elusive.

And speaking of `util.h`, those macros are provided for convenience. Most of those are to hide the implementation details similar to function calls (e.g. like `MAX()` or `MIN()`). They can be turned into static inline functions and still act the same. However, your proposed changes ARE NOT hiding implementation details, but hiding the declarations. They are not in the same category.

> * Open discussions on minimizing userspace impact on users that do not require it, for example, by not parsing almost all headers twice and autogenerating code for nothing (I want to believe this would not be accepted today).

Like I said above, until there is a full solution to be discussed, this is just hand waving at this point and nothing substantial could be discussed. As this is not only about changing the headers, but all code and build steps.

**Comment by dcpleung** [comment]
> > Yet the answer to the question of how hiding function declarations would be beneficial to users of those functions is still elusive.
> > And speaking of `util.h`, those macros are provided for convenience. Most of those are to hide the implementation details similar to function calls (e.g. like `MAX()` or `MIN()`). They can be turned into static inline functions and still act the same. However, your proposed changes ARE NOT hiding implementation details, but hiding the declarations. They are not in the same category.
> > > * Open discussions on minimizing userspace impact on users that do not require it, for example, by not parsing almost all headers twice and autogenerating code for nothing (I want to believe this would not be accepted today).
> > 
> > 
> > Like I said above, until there is a full solution to be discussed, this is just hand waving at this point and nothing substantial could be discussed. As this is not only about changing the headers, but all code and build steps.
> 
> I am not sure if it will be a good use of my time after seeing the initial reaction. Is anyone open to even discussing it?

What I oppose is on hiding the function declarations, via the macro you proposed. If you can find another solution, we can discuss the merit of that approach separately.

There could be a few ways of going forward:

1. Cleaning up the header by separating the `z_impl_` functions into their own headers. TBH, we should not be mixing private/internal implementations in public headers. The users of those APIs do not usually need to care about the implementations, but they would need to know the function arguments, their types, and the type and meaning of return values (if not void functions). This would de-clutter the public header files and make those files cleaner. This can be done separately.

2. The generated syscall related files could be pre-generated and checked into the tree. This avoids the cost involved during build, and can bake in the non-userspace shortcuts if needed (may even be just a `FUNC_ALIAS()` to `z_impl_`). However, this puts more burden on the driver and subsystem developers and maintainers as they would need to run those generation scripts every time the syscall function signatures change. This is cumbersome during development of new drivers and subsystems. Not to mention the argument against checking in generated code as they can be generated during build... so why check them in...

3. This is a variation of number 2 but only for downstream SDKs. For example, a company creating their own SDK to support their hardware can opt to use pre-generated bits, and distributes these bits with the SDK. The Zephyr main repo would still generate those code at build time. This, of course, assumes the users of said SDKs would not create their own syscalls. Though this could be guarded with a kconfig to revert to the generate-at-build-time behavior. This could also be useful to loadable module (LLEXT) development kit (EDK?) as it would lock down the supported syscalls that can be utilized by the modules (especially important as this would lock down the syscall dispatch table and avoiding changes to syscall enum values). The bits to enable this could be checked into the Zephyr main tree, but testing it in the public CI may require some massaging of test case definitions.

Feel free to propose another solution.

**Pull request closed without merging**